accepted/future-releases/enhanced-enums/feature-specification.md

```diff
@@ -346,8 +346,8 @@ enum Complex<T extends Pattern> with EnumComparable<Complex> implements Pattern
   }
 
   // Named constructor. Redirecting.
-  const Complex.captured(String regexpPattern)
-      : this("($regexpPattern)", RegExp);
+  const Complex.captured(String regexpPattern, T Function(String) factory)
+      : this("($regexpPattern)", factory);
```

Comment:
I think that  it was intended to have here `RegExp.new` instead of `RegExp`. But it'd be a compile-time error
`The argument type 'RegExp Function(String, {bool caseSensitive, bool dotAll, bool multiLine, bool unicode})' can't be assigned to the parameter type 'T Function(String)'.`
That's why I had to add additional argument `factory`

Reply:
See https://github.com/dart-lang/co19/blob/master/LanguageFeatures/Enhanced-Enum/example_A01_t01.dart

Reply:
Ah, of course: We could invoke the constructor with arbitrary type arguments, e.g., `myValue<Never>.captured('.*')`, and `RegExp` would only work with the value of `T` where `RegExp <: T`.

We could make it `RegExp.new as T Function()`, to preserve the idea that this constructor is meant to create `Complex` values based on `RegExp.new` as the factory (and not any other factory).

But it does illustrate that we very easily get into situations with `enum`s where the ability to pass a type argument (without additional constraints) to each constructor invocation is too flexible: It's never the _intention_ that we invoke this constructor with a type argument which is anything other than `RegExp`. So in this case it is inconvenient that we can't create instances in a static method (because a static method wouldn't have to take any type arguments, and it could return `Complex<RegExp>`), and that's because the result from such a call would not be constant.

@lrhn, I think it would make sense to mention this dilemma &mdash; unless you have a different design that actually eliminates the dilemma. ;-)

Reply:
It should have been `RegExp.new`. I consistently forget that you have to write `.new` to tear off the constructor.
And then it indeed doesn't work because we can't specialize the constructor. The `factory` argument is better, because then the `T` type argument can be inferred at from the argument at instantionation.